datatypes/src/raster/raster.rs

```diff
@@ -0,0 +1,274 @@
+        ensure!(
+            grid_dimension.capacity() != data_container.capacity(),
+            error::DimnsionCapacityDoesNotMatchDataCapacity {
+                dimension_cap: grid_dimension.capacity(),
+                data_cap: data_container.capacity()
+            }
+        );
+
+        Ok(Self {
+            grid_dimension,
+            data_container,
+            no_data_value,
+            temporal_bounds,
+            geo_transform,
+        })
+    }
+}
+
+impl<D, T, C> TemporalBounded for BaseRaster<D, T, C> {
+    fn temporal_bounds(&self) -> TimeInterval {
+        self.temporal_bounds.clone()
+    }
+}
+
+impl<D, C, T> SpatialBounded for BaseRaster<D, C, T>
+where
+    D: GridDimension,
+{
+    fn spatial_bounds(&self) -> BoundingBox2D {
+        let top_left_coord = self.geo_transform.grid_2d_to_coordinate_2d((0, 0));
+        let lower_right_coord = self.geo_transform.grid_2d_to_coordinate_2d((
+            self.grid_dimension.size_of_y_axis(),
+            self.grid_dimension.size_of_x_axis(),
+        ));
+        BoundingBox2D::new_ul_lr_unchecked(top_left_coord, lower_right_coord)
+    }
+}
+
+impl<D, T, C> Raster<D, T, C> for BaseRaster<D, T, C>
+where
+    D: GridDimension,
+    T: Copy,
+    C: Capacity,
+{
+    fn dimension(&self) -> &D {
+        &self.grid_dimension
+    }
+    fn no_data_value(&self) -> Option<T> {
+        self.no_data_value
+    }
+    fn data_container(&self) -> &C {
+        &self.data_container
+    }
+    fn geo_transform(&self) -> &GeoTransform {
+        &self.geo_transform
+    }
+}
+
+pub trait GridPixelAccess<T, I> {
+    /// Gets the value at a pixels location
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geoengine_datatypes::raster::{Raster, Dim, SimpleRaster2d, TimeInterval, GridPixelAccess};
+    ///
+    /// let mut raster2d = SimpleRaster2d::new(
+    ///    [3, 2].into(),
+    ///    vec![1,2,3,4,5,6],
+    ///    None,
+    ///    TimeInterval::default(),
+    ///    [1.0, 1.0, 0.0, 1.0, 0.0, 1.0].into(),
+    /// ).unwrap();
+    /// let value = raster2d.pixel_value_at_grid_index(&(1, 1)).unwrap();
+    /// assert_eq!(value, 4);
+    /// ```
+    ///
+    fn pixel_value_at_grid_index(&self, grid_index: &I) -> Result<T>;
+}
+
+pub trait GridPixelAccessMut<T, I> {
+    /// Sets the value at a pixels location
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geoengine_datatypes::raster::{Raster, Dim, SimpleRaster2d, TimeInterval, GridPixelAccessMut};
+    ///
+    /// let mut raster2d = SimpleRaster2d::new(
+    ///    [3, 2].into(),
+    ///    vec![1,2,3,4,5,6],
+    ///    None,
+    ///    TimeInterval::default(),
+    ///    [1.0, 1.0, 0.0, 1.0, 0.0, 1.0].into(),
+    /// ).unwrap();
+    /// raster2d.set_pixel_value_at_grid_index(&(1, 1), 9).unwrap();
+    /// assert_eq!(raster2d.data_container(), &[1,2,3,9,5,6]);
+    /// ```
+    ///
+    fn set_pixel_value_at_grid_index(&mut self, grid_index: &I, value: T) -> Result<()>;
```

Comment:
Is immutability an important constraint for our types? If yes, we should do something builder like with a mutable grid.

Reply:
i guess we have to discuss this